Initial Edilkamin stove integration

AndreMiras opened the pull request:
## Proposed change

This is a first iteration with a basic set of feature:
- turn on/off
- get/set target temperature
- get current temperature

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

The bluetooth is currently only used for device discovery. The device is then driven from a cloud based API.
It seems like alternatively the device can also be driven fully from bluetooth, but the underlying `edilkamin` Python library doesn't support it yet.


## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository] https://github.com/home-assistant/home-assistant.io/pull/25072

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly. Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`. Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

AndreMiras committed:
Initial Edilkamin stove integration

This is a first iteration with a basic set of feature:
- turn on/off
- get/set target temperature
- get current temperature

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

The bluetooth is currently only used for device discovery.
The device is then driven from a cloud based API.
It seems like alternatively the device can also be driven fully from
bluetooth, but the underlying `edilkamin` Python library doesn't support
it yet.

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
  https://github.com/home-assistant/home-assistant.io/pull/25072

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

onorbe commented:
im very interested in this!!!

akx commented on homeassistant/components/edilkamin/entity.py:
```diff
@@ -0,0 +1,75 @@
+    edilkamin.Power.OFF: HVACMode.OFF,
+    edilkamin.Power.ON: HVACMode.HEAT,
+}
+hvac_mode_to_power = {
+    HVACMode.OFF: edilkamin.Power.OFF,
+    HVACMode.HEAT: edilkamin.Power.ON,
+}
```
You could just
```suggestion
hvac_mode_to_power = {hvac: power for (power, hvac) in power_to_hvac.items()}
```

AndreMiras replied:
Addressed

akx reviewed (commented):
> The bluetooth is used for the device discovery and we're using the simplepyble module doesn't currently provide pre-built wheels for Linux ARM (e.g. Raspberry). In the future we may provide multiple optional bluetooth backends such as pybluez.

I'm afraid using another library this will conflict _hard_ with how Home Assistant does Bluetooth internally using `bleak`.
If you want to use the HA internal Bluetooth bits (which do e.g. device discovery for you and automagically allow showing new matching devices as ready-to-set-up in the HA UI and so on), the `edilkamin` library should be agnostic as to how the data is sent or received to the device, but just parse it.

You might want to look at the other `_ble` integrations for inspiration (though many of them are sensor only, and I'm not sure how hooking up the active command bits would work; cc @bdraco on that...).

**EDIT:** if https://github.com/AndreMiras/edilkamin.py/blob/5e52900ef938391b8a59416da60516308cfdc7f1/edilkamin/api.py#L41-L59 is indeed the only thing the `edilkamin` library needs BLE for, it should maybe be an optional feature (pip extra) with a late import for the BLE library, so you could use the HA native Bluetooth discovery for MACs here, and people who want to build on `edilkamin` without HA could do e.g. `pip install edilkamin[ble]`. Please feel free to ping me if you need help with the pip extras bits :)

shisva commented:
> > The bluetooth is used for the device discovery and we're using the simplepyble module doesn't currently provide pre-built wheels for Linux ARM (e.g. Raspberry). In the future we may provide multiple optional bluetooth backends such as pybluez.
> 
> I'm afraid using another library this will conflict _hard_ with how Home Assistant does Bluetooth internally using `bleak`. If you want to use the HA internal Bluetooth bits (which do e.g. device discovery for you and automagically allow showing new matching devices as ready-to-set-up in the HA UI and so on), the `edilkamin` library should be agnostic as to how the data is sent or received to the device, but just parse it.
> 
> You might want to look at the other `_ble` integrations for inspiration (though many of them are sensor only, and I'm not sure how hooking up the active command bits would work; cc @bdraco on that...).
> 
> **EDIT:** if https://github.com/AndreMiras/edilkamin.py/blob/5e52900ef938391b8a59416da60516308cfdc7f1/edilkamin/api.py#L41-L59 is indeed the only thing the `edilkamin` library needs BLE for, it should maybe be an optional feature (pip extra) with a late import for the BLE library, so you could use the HA native Bluetooth discovery for MACs here, and people who want to build on `edilkamin` without HA could do e.g. `pip install edilkamin[ble]`. Please feel free to ping me if you need help with the pip extras bits :)

Hi, I’m using https://github.com/home-assistant/core/tree/dev/homeassistant/components/yalexs_ble to control my lock via Bluetooth and it is working pretty smooth

AndreMiras commented:
Hi @akx,
Thanks a mil for the valuable feedback. I've tried to ping you on Discord (`Aarnink#1056`) by the way, but I don't know if you're active there.


> I'm afraid using another library this will conflict _hard_ with how Home Assistant does Bluetooth internally using `bleak`. If you want to use the HA internal Bluetooth bits (which do e.g. device discovery for you and automagically allow showing new matching devices as ready-to-set-up in the HA UI and so on), the `edilkamin` library should be agnostic as to how the data is sent or received to the device, but just parse it.

Good to know, I thought it might be the case indeed. I've refactored the code to leverage `async_discovered_service_info()` and use the edilkamin library to only parse the result.

> You might want to look at the other `_ble` integrations for inspiration (though many of them are sensor only, and I'm not sure how hooking up the active command bits would work; cc @bdraco on that...).

I did look a lot of that indeed, but had hard time finding a cloud based integration that was only using bluetooth for the initial device discovery like in my case.
Basically that device is driven from an API on the "cloud", but no endpoint would return the device "ID", that can only be extracted using the device bluetooth MAC at the moment.
So it might look like a bluetooth integration, but it's not.
In the future I may also reverse their bluetooth protocol and make a fully bluetooth alternative integration, but for now I'm using their cloud integration.

> https://github.com/AndreMiras/edilkamin.py/blob/5e52900ef938391b8a59416da60516308cfdc7f1/edilkamin/api.py#L41-L59 is indeed the only thing the `edilkamin` library needs BLE for, it should maybe be an optional feature (pip extra) with a late import for the BLE library, so you could use the HA native Bluetooth discovery for MACs here, and people who want to build on `edilkamin` without HA could do e.g. `pip install edilkamin[ble]`. Please feel free to ping me if you need help with the pip extras bits :)

Addressed in the last release:
https://github.com/AndreMiras/edilkamin.py/blob/1.1.0/setup.py#L36

AndreMiras commented:
> Hi, I’m using https://github.com/home-assistant/core/tree/dev/homeassistant/components/yalexs_ble to control my lock via Bluetooth and it is working pretty smooth

Hi @shisva thanks for sharing.
This integration doesn't actually rely on bluetooth to control the device.
The device is connected to the cloud and you control them by calling a cloud API.
The bluetooth is only used for the initial discovery in order to know the ID of the device you're controlling

akx commented:
Hi @AndreMiras, you're welcome for the feedback! Great you got things moving forward, the changes to the API package look just like I imagined they'd be. I'm not very (read "at all really") active on Discord, but for future reference my handle there is akx#0725, so I'm not sure who you tried to reach 😁

AndreMiras commented:
Hi @akx,
Anything I could do on my side to help with the review/merge?

akx commented:
Hi @AndreMiras, I'm not sure I can do much here, I'm just a contributor doing reviews "on the side". A core developer would probably need to take a look at this too.

tggm commented on homeassistant/components/edilkamin/config_flow.py:
```diff
@@ -0,0 +1,98 @@
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
```
Use `cv.str` instead of `str`. Import from:

`import homeassistant.helpers.config_validation as cv`

AndreMiras replied:
Addressed, I didn't know.
I'm wondering why the scaffolding script doesn't use it :man_shrugging: 
Do we know?
https://github.com/home-assistant/core/blob/2022.12.1/script/scaffold/templates/config_flow/integration/config_flow.py#L21-L23

AndreMiras commented on homeassistant/components/edilkamin/entity.py:
```diff
@@ -0,0 +1,81 @@
+)
+from homeassistant.const import ATTR_TEMPERATURE, TEMP_CELSIUS
+
+_LOGGER = logging.getLogger(__name__)
```
Oh right I missed that one module indeed, good catch thanks

ElBoonoS commented:
This looks like a great integration; hats the next step here? Anything the community can do ?

AndreMiras commented:
Rebased and fixed conflicts

AndreMiras commented:
Rebased to dev

edenhaus commented on homeassistant/components/edilkamin/translations/en.json:
```diff
@@ -0,0 +1,21 @@
+{
```
Please remove the folder `translations`

AndreMiras replied:
Addressed

edenhaus commented on homeassistant/components/edilkamin/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/edilkamin",
+  "requirements": ["edilkamin==1.1.0"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
```
Please remove empty keys

AndreMiras replied:
Addressed

edenhaus reviewed (changes requested):
Please make sure CI is passing

edenhaus reviewed (changes requested):
The integration tests are failing, please fix them

AndreMiras commented:
I'm not too sure why the CI didn't trigger this time, I wated for a few days to see if it would start, but it didn't.
So re-requested for review, I've checked as much as I knew in local

frenck commented:
> I'm not too sure why the CI didn't trigger this time

As you are a first-time contributor and fairly new to GitHub, it means you trigger a security mechanism. Any change needs manual approval to run on CI right now. This will get better over time for your account.

../Frenck

AndreMiras commented:
> > I'm not too sure why the CI didn't trigger this time
> 
> As you are a first-time contributor and fairly new to GitHub, it means you trigger a security mechanism. Any change needs manual approval to run on CI right now. This will get better over time for your account.
> 
> ../Frenck

OK thanks, could you trigger a build?
I've rebased and added missing bluetooth auto mock which should fix the build

AndreMiras commented:
I've just updated the branch again fixing the test coverage it should be a 100% again now.
@frenck could you run the build again to see if I'm missing anything?

AndreMiras commented:
Tests look better now. Except that `tests/components/homeassistant_sky_connect/test_config_flow.py` one which is also failing in the dev branch (which I rebased on).

emontnemery commented:
(Close / open to rerun CI)

emontnemery commented on homeassistant/components/edilkamin/climate.py:
```diff
@@ -0,0 +1,32 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the stove with config flow."""
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    ble_devices = tuple(
+        {"name": discovery_info.name, "address": discovery_info.address}
+        for discovery_info in async_discovered_service_info(hass, False)
+    )
+    mac_addresses = edilkamin.discover_devices_helper(ble_devices)
+    entities = [
+        EdilkaminEntity(username, password, mac_address)
+        for mac_address in mac_addresses
+    ]
```
This is essentially doing bluetooth discovery in the climate platform, that's really strange.

Understood that the `edilkamin` supporting library is not BLE based, but:
- Let's go the standard route of one device per config entry, the config flow should discover devices and offer the user which one to pick
- What happens if the user does not have a bluetooth adapter, does `edilkamin.discover_devices_helper` not work then (that's a very odd cloud API in that case..)
- Can we add bluetooth discovery?

emontnemery commented on homeassistant/components/edilkamin/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+        self.hass = hass
+
+    async def authenticate(self, username: str, password: str) -> str:
+        """Authenticate with the host and return the token or raise an exception."""
```
Are we authenticating with a local host or with some cloud API?

emontnemery commented on homeassistant/components/edilkamin/entity.py:
```diff
@@ -0,0 +1,80 @@
+hvac_mode_to_power = {hvac: power for (power, hvac) in power_to_hvac.items()}
+
+
+class EdilkaminEntity(ClimateEntity):
```
Please move this to `climate.py`. An `entity.py` is OK if it contains a base class, shared by several kind of entities.

emontnemery replied:
Please look into the [updated entity naming rules](https://developers.home-assistant.io/docs/core/entity#entity-naming) and set `has_entity_name` to `True`

emontnemery commented on homeassistant/components/edilkamin/entity.py:
```diff
@@ -0,0 +1,80 @@
+        - the username/password to login
+        - the MAC address that identifies the stove
+        """
+        self._attr_name = f"Stove ({mac_address})"
```
Doesn't the cloud API provide something better than a mac address?

emontnemery commented on homeassistant/components/edilkamin/entity.py:
```diff
@@ -0,0 +1,80 @@
+        self._attr_unique_id = mac_address
+        self._username = username
+        self._password = password
+        self._mac_address = mac_address
```
Delete this, it's already stored in `_attr_unique_id`
```suggestion
```

emontnemery commented on homeassistant/components/edilkamin/entity.py:
```diff
@@ -0,0 +1,80 @@
+        """Login to refresh the token."""
+        return cast(str, edilkamin.sign_in(self._username, self._password))
+
+    def update(self) -> None:
```
I strongly suggest using a [`DataUpdateCoordinator`](https://developers.home-assistant.io/docs/integration_fetching_data/#coordinated-single-api-poll-for-data-for-all-entities) instead
